influunt-api/modules/influunt-core/app/engine/GerenciadorDeEstagios.java

```diff
@@ -0,0 +1,287 @@
+
+            estagioPlanoAnterior = estagioPlanoAtual;
+            estagioPlanoAtual = estagioPlano;
+        }
+    }
+
+    private void executaAgendamentoTrocaDePlano() {
+        agendamento.setMomentoDaTroca(tempoDecorrido);
+        callback.onTrocaDePlanoEfetiva(agendamento);
+        reconhecePlano(this.agendamento.getPlano());
+        this.agendamento = null;
+    }
+
+    public void trocarPlano(AgendamentoTrocaPlano agendamentoTrocaPlano) {
+        agendamentoTrocaPlano.setMomentoPedidoTroca(tempoDecorrido);
+        agendamentoTrocaPlano.setAnel(anel);
+        agendamento = agendamentoTrocaPlano;
+    }
+
+    private void reconhecePlano(Plano plano) {
+        reconhecePlano(plano, false);
+    }
+
+    private void reconhecePlano(Plano plano, boolean inicio) {
+        if (this.plano != null) {
+            modoAnterior = this.plano.getModoOperacao();
+        }
+
+        this.plano = plano;
+        this.tabelaDeTemposEntreVerde = this.plano.tabelaEntreVerde();
+        this.listaOriginalEstagioPlanos = this.plano.ordenarEstagiosPorPosicaoSemEstagioDispensavel();
+        this.listaEstagioPlanos = new ArrayList<>(listaOriginalEstagioPlanos);
+        contadorEstagio = 0;
+        contadorIntervalo = 0L;
+        contadorDeCiclos = 0L;
+
+
+        if (inicio && listaEstagioPlanos.size() > 0) {
+            this.estagioPlanoAtual = listaEstagioPlanos.get(listaEstagioPlanos.size() - 1);
+        }
+
+
+        geraIntervalos(0);
+
+        if (!inicio) {
+            EventoMotor eventoMotor = new EventoMotor(null, TipoEvento.TROCA_DE_PLANO_NO_ANEL, agendamento.getPlano().getPosicao(), agendamento.getAnel(), agendamento.getMomentoOriginal(), agendamento.getMomentoDaTroca());
+            this.intervalos.get(0L).addEvento(contadorIntervalo, eventoMotor);
+        }
+    }
+
+    private void geraIntervalos(Integer index) {
+        GeradorDeIntervalos gerador = GeradorDeIntervalos.getInstance(this.intervalos, this.plano,
+                this.modoAnterior, this.listaEstagioPlanos,
+                this.estagioPlanoAtual, this.tabelaDeTemposEntreVerde,
+                index);
+
+        Pair<Integer, RangeMap<Long, IntervaloEstagio>> resultado = gerador.gerar(index);
+
+        this.contadorEstagio += resultado.getFirst();
+        this.intervalos = resultado.getSecond();
+
+        modoAnterior = this.plano.getModoOperacao();
+    }
+
+    @Override
+    public void onEvento(EventoMotor eventoMotor) {
+        long offset = 0;
+        if (this.intervalos.get(contadorIntervalo) == null) {
+            offset = contadorIntervalo - 1;
+        }
+        this.intervalos.get(offset).addEvento(contadorIntervalo, eventoMotor);
+        GerenciadorDeEventos.onEvento(this, eventoMotor);
+    }
+
+    public void atualizaEstagiosCicloAtual(EstagioPlano estagioPlano) {
+        List<EstagioPlano> novaLista = new ArrayList<>();
+        final Boolean[] adicionado = {false};
```

Comment:
Senão não pode ser alterado dentro de stream, se for uma variavel simples teremos que usar for dentro de for.... acho melhor assim, mais se preferirem alteramos para for dentro de for o que acha @rodrigosol e @ppires??